### ISSUE TITLE
Proposal: Symbol at point as default search pattern for `counsel-*g` and `counsel-git-grep` and all other search tools

### ISSUE THREAD

**Issue by Alexander-Shukaev** [abo-abo/swiper#1867]
Would it be possible to input `symbol-at-point` by default whenever a `counsel`-based search tool is invoked?  That is try the stock `grep-recursive`.  I think it's very cleverly implemented in that it displays `symbol-at-point` in parentheses as a default pattern, so if you hit `RET` without typing anything, this will become your pattern.  Similar feature could be implemented for `counsel`-based search tools.  That is be default the input field remains empty as it is now but there would be `symbol-at-point` in parentheses displayed in a similar manner to `grep-recursive` and the invoked search tool would immediately search for that default pattern already.  In other words, you could immediately trigger `ivy-occur` without even typing anything because there are matches already.  However, if you start typing anything, then your pattern of course overtakes, while as soon as you delete your input completely, again the default pattern should kick in.  To be reasonable, a `defcustom` to enable this behavior should be introduced.

**Comment by mookid** [comment]
agreed. I have some code that does works around that limitation in my config for a while now.

**Comment by vishesh** [comment]
This is actually useful as reasonable number of times I wish to search what's under the cursor. I borrowed this[1] which does exactly same, however the annoying thing about this is that its not uncommon to search something you are not under, in which case you have manually clear the input  and that I've found to be more annoying. Perhaps that problem can be mitigated to some extent by keeping input selected?

[1]  https://github.com/seagle0128/.emacs.d/blob/master/lisp/init-ivy.el#L299

**Comment by Alexander-Shukaev** [comment]
Indeed, this requires slightly more effort to implement than just inject initial input (which is unacceptable due to annoyance of clearing as you already mentioned).

**Comment by basil-conto** [comment]
@Alexander-Shukaev The closest existing solution to what you described is specifying the symbol at point as the `DEF` keyword/argument to `ivy-read`/`completing-read`. IIRC, Ivy currently adds such default items to the list of candidates unless already present, and preselects them. AIUI, the only difference between this and what you describe is that the prompt does not automatically change depending on whether a `DEF` argument was specified. By convention, it's up to the caller to modify the prompt accordingly, so I'm not sure we could successfully be smart about this in ivy without becoming intrusive. Besides, the reason `(default: foo)` is needed in vanilla Emacs completion is because there is no realtime visualisation of the candidate list, whereas Ivy makes it clear that the default is preselected.

**Comment by Alexander-Shukaev** [comment]
That's fine, I can live with that, whatever is "native" for the framework should stay so.  Then, all what's left to implement this now is to add  `:def ...` all over the place?

**Comment by Alexander-Shukaev** [comment]
I suppose to be more flexible, what could be done is a variable similar to `ivy-initial-inputs-alist`, e.g. `ivy-read-def-alist`, where one could add whatever they want in a custom fashion.  I'd appreciate a `symbol-at-point` for search commands as a default example though.  On a side note, why `ivy-initial-inputs-alist` is a `defvar` rather than `defcustom`?

**Comment by basil-conto** [comment]
> I suppose to be more flexible, what could be done is a variable similar to `ivy-initial-inputs-alist`, e.g. `ivy-read-def-alist`, where one could add whatever they want in a custom fashion. I'd appreciate a `symbol-at-point` for search commands as a default example though.

SGTM. The values of the alist would need to optionally be functions so as to determine the symbol at point dynamically.

>  On a side note, why `ivy-initial-inputs-alist` is a `defvar` rather than `defcustom`?

I don't know; @abo-abo might. I'm guessing oversight / historical accident.
